orm/model/ormlist/options.go

```diff
@@ -0,0 +1,77 @@
+// Package ormlist defines options for listing items from ORM indexes.
+package ormlist
+
+import (
+	"github.com/cosmos/cosmos-sdk/orm/encoding/encodeutil"
+	"github.com/cosmos/cosmos-sdk/orm/internal/listinternal"
+)
+
+// Option represents a list option.
+type Option = listinternal.Option
+
+// Start defines the values to use to start range iteration. It cannot be
+// combined with Prefix.
+//
+// Values must correspond in type to the index's fields and the number of values
+// provided cannot exceed the number of fields in the index, although fewer
+// values can be provided.
+//
+// Range iteration can only be done for start and end values which are
+// well-ordered, meaning that any unordered components must be equal. Ex.
+// the bytes type is considered unordered, so a range iterator is created
+// over an index with a bytes field, both start and end must have the same
+// value for bytes.
+func Start(values ...interface{}) Option {
```

Comment:
If I understood this correctly, this is a lot ordering unsafe, although ORM would error, I'd personally prefer:

```
type ListValue struct {
   Name protoreflect.FullName
   Value interface{} // or even protoreflect.Value for max safety.
}
```

Reply:
Yeah I mean this is the hardest part of the API to get right for end users and probably where codegen is needed most... though I'm not sure how we'd deal with partial prefixes/range queries even with codegen

Reply:
Also just specifying field name doesn't guarantee correct type. Using int vs uint32 for instance will cause an error. And this is more verbose and requires wrapping in protoreflect.Name...

How hard would codegen for this be?

Reply:
Mh.. not sure tbh.

To have max type safety, we'd want (ideally):

given Index, of object `Foo`:
```
A: int64
B: string
```

```go

// codegenned

type FooLister interface { 
 IndexIdentifier() uint64
 Values() []ormlist.ListValue
 fooOnly() // gives type safety in a way that in the client for object Foo only the generated implementers can be used.
}

// AB identifies the index name
type ABListOption struct {
   A *int64
   B *string
}

func (o *ABListOption) IndexIdentifier() { index-id }
func (o *ABListOption) Values() []ormlist.ListValue {
   var v []ormlist.ListValue
   if o.A != nil {v = append(v, protoreflect.ValueOfInt(o)}
   ... etc
   return v
}

func (a *ABORMClient) List(ctx context.Context, listOpt FooLister) (FooIterator, error) {
   indexId := listOpt.IndexIdentifier()
   index := a.orm.GetIndex(indexId)
   iter := index.Iter(listOpt.Values())
   return footer{iter:iter}, nil
}

// codegenned end
// this is in ormlist pkg

type ListValue struct {
   Value protoreflect.Value
   Field protoreflect.FullName
}

Reply:
Generating the full combo of possible full and partial indexes IMO is madness, ORM generated code APIs would be giga messy.

Instead, one List method which accepts multiple type safe ListOptions which are index specific and can be partially filled in a type safe way might be ideal.

Reply:
Totally generating all the combos will be messy but I think there are ways...

This current API is the best I could come up with for dynamic usage after trying a few things. Usage in general will be like this where you name the index variable according to the fields so you don't forget:
```go
abIndex := fooTable.GetIndex("a,b")
abIndex.Get(int64(1), "abc")
```
I think that's not too bad considering what other go libraries like gorm do and tests will catch any type problems.

---

Regarding codegen, you could have just have two constructors for to handle this `ABListOption`:
```go
func AListOption(a int64) FooLister { ... }

func ABListOption(a int64, b string) FooLister {...}
```

I think that's a pretty straightforward way to handle all the permutations rather than pointers because for `*int64` you can't just write `&1` in go.

Reply:
My sense is that the current API is the least painful way to use a table without codegen (after having tried to play around with it a bit). Would it be okay to go with this for now and address type safety with a good codegen design @fdymylja ?

Reply:
Discussed in discord and it SGTM.